Ada Examples/Examples Based On OpenGL Superbible/Simplest/src/initialization.adb

```diff
@@ -0,0 +1,123 @@
+
+        Window_Width       : Interfaces.C.Int := 800;
+        Window_Height      : Interfaces.C.Int := 600;
+        Cursor                    : Glfw.Input.Mouse.Cursor_Mode := Glfw.Input.Mouse.Hidden;
+        Can_Start               : Boolean := False;
+    begin
+        Set_Window_Hints;
+        Main_Window.Init(Window_Width, Window_Height, "OpenGL SuperBible Example");
+        glfw.Windows.Context.Make_Current(Main_Window'Access);
+        Enable_Callbacks(Main_Window);
+
+        gl.Toggles.Enable(gl.Toggles.Depth_Test);
+        gl.Buffers.Set_Depth_Function(gl.Types.Less);
+        Main_Window.Set_Cursor_Mode(Cursor);
+        Utilities.Show_GL_Data;
+
+        Can_Start := Startup(Main_Window, Rendering_Program, Vertex_Array);
+        return Can_Start;
+
+ exception
+      when anError : Constraint_Error =>
+         Put("Initialize returned constraint error: ");
+         Put_Line(Exception_Information(anError));
+        return Can_Start;
+
+      when anError : GL.Errors.Invalid_Operation_Error =>
+         Put_Line("Initialize returned an invalid operation error: ");
+         Put_Line(Exception_Information(anError));
+        return Can_Start;
+
+      when anError :  others =>
+         Put_Line("An exceptiom occurred in Initialize.");
+         Put_Line(Exception_Information(anError));
+        return Can_Start;
+
+    end Initialize;
+
+    --  ------------------------------------------------------------------------------------------------------------------------
+
+   procedure Set_Window_Hints is
+      Min_Major_Version    : Integer := 3;
+      Minor_Version            : Integer := 2;
+   begin
+        glfw.Windows.Hints.Set_Minimum_OpenGL_Version(Min_Major_Version, Minor_Version);
+        glfw.Windows.Hints.Set_Forward_Compat(True);
+        glfw.Windows.Hints.Set_Profile(glfw.Windows.Context.Core_Profile);
+        glfw.Windows.Hints.Set_Debug_Context(True);
+        -- Set samples to 16 before taking screen shots.
+        glfw.Windows.Hints.Set_Samples(4);
+
+    end Set_Window_Hints;
+
+    --  ------------------------------------------------------------------------------------------------------------------------
+
+    function Startup(Main_Window : Window_Types.tWindow;
+                     Rendering_Program   : out GL.Objects.Programs.Program;
+                     Vertex_Array              : in out GL.Objects.Vertex_Arrays.Vertex_Array_Object)
+                     return Boolean is
+
+        OK  : Boolean := Shaders_Program.Make_Shader_Program(Main_Window, Rendering_Program);
+    begin
+        if OK then
+           Vertex_Array.Initialize_Id;
+           Vertex_Array.Bind;
+            gl.Toggles.Enable(gl.Toggles.Vertex_Program_Point_Size);
+        else
+           Put_Line("Shader program linking failed.");
+           Put_Line("Log:");
+           Put_Line(Rendering_Program.Info_Log);
+        end if;
+        Utilities.Show_Shader_Program_Data(Rendering_Program);
+        return OK;
```

Comment:
Just so you are aware, you can do:

```ada
return OK : Boolean := Shaders_Program.Make_Shader_Program (Main_Window, Rendering_Program) do
   if OK then
      Vertex_Array.Initialize_Id;
      Vertex_Array.Bind;
      GL.Toggles.Enable (GL.Toggles.Vertex_Program_Point_Size);
   else
      Put_Line ("Shader program linking failed.");
      Put_Line ("Log:");
      Put_Line (Rendering_Program.Info_Log);
   end if;
   Utilities.Show_Shader_Program_Data (Rendering_Program);
end return;
```

The idea is to set the return value but afterwards to something depending on it before actually returning.